File: python/circular_queue.py
class CircularQueue:
    def __init__(self, size):
        self.size = size
        self._queue = [None] * size
        self.head = -1
        self.tail = -1

    def is_empty(self):
        return (self.head == -1) and (self.tail == -1)

    def is_full(self):
        return (self.tail + 1) % self.size == self.head

    def enqueue(self, new_value):
        if self.is_full():
            raise FullCircularQueueError()

        # adjust indexes for head and tail
        if self.is_empty():
            self.head = 0
            self.tail = 0
        
        else:
            self.tail = (self.tail + 1) % self.size
        
        # add the new value to the queue
        self._queue[self.tail] = new_value

    def dequeue(self):
        if self.is_empty():
            raise EmptyCircularQueueError()

        ret = self._queue[self.head]

        # if we are dequeueing the last item from the queue, then put into empty state
        if self.head == self.tail:
            self.head = -1
            self.tail = -1

        # otherwise, adjust the head index normally
        else:
            self.head = (self.head + 1) % self.size

        return ret

    def peek(self):
        if self.is_empty():
            raise EmptyCircularQueueError()

        return self._queue[self.head]

    def display(self):
        display_value = self._get_display_value()
        print(display_value)

    def _get_display_value(self):
        if self.is_empty():
            return []
        
        if self.head <= self.tail:
            start = self.head
            end = self.tail + 1
            return self._queue[start : end]

        else:
            index = self.head
            first_list = self._queue[index:]

            index = self.tail + 1
            second_list = self._queue[:index]

            return first_list + second_list

class EmptyCircularQueueError(Exception):
    def __init__(self, message="Circular Queue is empty!"):
        super().__init__(message)

class FullCircularQueueError(Exception):
    def __init__(self, message="Circular Queue is full!"):
        super().__init__(message)

File: javascript/circular_queue.js
class CircularQueue {
    constructor(size) {
        this._size = size;
        this._queue = new Array(size);
        this._head = -1;
        this._tail = -1;
    }

    isEmpty() {
        return (this._head === -1) && (this._tail === -1);
    }

    isFull() {
        return (this._tail + 1) % this._size === this._head;
    }

    enqueue(newValue) {
        if (this.isFull()) {
            throw new FullCircularQueueError();
        }

        // adjust indexes for head and tail
        if (this.isEmpty()) {
            this._head = 0;
            this._tail = 0;
        }

        else {
            this._tail = (this._tail + 1) % this._size;
        }

        // add the new value to the queue
        this._queue[this._tail] = newValue;
    }

    dequeue() {
        if (this.isEmpty()) {
            throw new EmptyCircularQueueError();
        }

        const ret = this._queue[this._head];

        // if we are dequeueing the last item from the queue, then put into empty state
        if (this._head === this._tail) {
            this._head = -1;
            this._tail = -1;
        }

        // otherwise, adjust the head index normally
        else {
            this._head = (this._head + 1) % this._size;
        }

        return ret;
    }

    peek() {
        if (this.isEmpty()) {
            throw new EmptyCircularQueueError();
        }

        return this._queue[this._head];
    }

    display() {
        const displayValue = this._getDisplayValue();
        console.log(displayValue);
    }

    _getDisplayValue() {
        if (this.isEmpty()) {
            return [];
        }

        if (this._head <= this._tail) {
            const start = this._head;
            const end = this._tail;
            return this._queue.slice(start, end);
        }

        else {
            const firstIndex = this._head;
            const firstArr = this._queue.slice(firstIndex);

            const secondIndex = this._tail + 1;
            const secondArr = this._queue.slice(0, secondIndex);

            return firstArr.concat(secondArr);
        }
    }
}

class EmptyCircularQueueError extends Error {
    constructor(message="Circular Queue is empty!") {
        super(message);
        this.name = "EmptyCircularQueueError";
    }
}

class FullCircularQueueError extends Error {
    constructor(message="Circular Queue is full!"){
        super(message);
        this.name = "FullCircularQueueError";
    }
}